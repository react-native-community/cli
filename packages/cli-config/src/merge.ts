import deepmerge from 'deepmerge';

/**
 * `deepmerge` concatenates arrays by default instead of overwriting them.
 * We define custom merging function for arrays to change that behaviour
 */
export default function merge<X, Y>(x: X, y: Y) {
  return deepmerge(x, y, {
    arrayMerge: (_destinationArray: any[], sourceArray: any[]): any[] =>
      sourceArray,
  });
}
