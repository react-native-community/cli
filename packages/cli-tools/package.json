{
  "name": "@react-native-community/cli-tools",
  "version": "20.0.1",
  "license": "MIT",
  "main": "build/index.js",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@vscode/sudo-prompt": "^9.0.0",
    "appdirsjs": "^1.2.4",
    "execa": "^5.0.0",
    "find-up": "^5.0.0",
    "launch-editor": "^2.9.1",
    "mime": "^2.4.1",
    "ora": "^5.4.1",
    "picocolors": "^1.1.1",
    "prompts": "^2.4.2",
    "semver": "^7.5.2"
  },
  "devDependencies": {
    "@react-native-community/cli-types": "20.0.1",
    "@types/mime": "^2.0.1",
    "@types/node": "^20.0.0",
    "@types/prompts": "^2.4.4",
    "@types/shell-quote": "^1.7.1"
  },
  "files": [
    "build",
    "!*.d.ts",
    "!*.map"
  ],
  "homepage": "https://github.com/react-native-community/cli/tree/main/packages/cli-tools",
  "repository": {
    "type": "git",
    "url": "https://github.com/react-native-community/cli.git",
    "directory": "packages/tools"
  },
  "gitHead": "3614a01613030c0a328ad2e8bb5d897e5bae1456"
}
