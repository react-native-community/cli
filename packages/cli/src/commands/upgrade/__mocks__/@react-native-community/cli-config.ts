export default function mockedLoadConfig() {
  return {
    root: '/project/root',
    reactNativePath: '',
    commands: [],
    platforms: {
      ios: {projectConfig: () => null, dependencyConfig: () => null},
      android: {projectConfig: () => null, dependencyConfig: () => null},
    },
    project: {
      ios: null,
      android: null,
    },
    dependencies: {},
    assets: [],
  };
}
