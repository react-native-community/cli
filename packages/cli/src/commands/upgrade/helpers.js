// @flow
import https from 'https';

export const fetch = (url: string) =>
  new Promise<string>((resolve, reject) => {
    const request = https.get(url, response => {
      if (response.statusCode < 200 || response.statusCode > 299) {
        reject(
          new Error(`Failed to load page, status code: ${response.statusCode}`)
        );
      }
      const body = [];
      response.on('data', chunk => body.push(chunk));
      response.on('end', () => resolve(body.join('')));
    });
    request.on('error', err => reject(err));
  });
