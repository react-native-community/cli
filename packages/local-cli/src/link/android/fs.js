/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

const fs = require('fs-extra');

exports.readFile = file => () => fs.readFileSync(file, 'utf8');

exports.writeFile = (file, content) =>
  content
    ? fs.writeFileSync(file, content, 'utf8')
    : c => fs.writeFileSync(file, c, 'utf8');
