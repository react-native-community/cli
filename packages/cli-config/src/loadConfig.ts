import path from 'path';
import {
  UserDependencyConfig,
  ProjectConfig,
  DependencyConfig,
  UserConfig,
  Config,
} from '@react-native-community/cli-types';
import {
  findProjectRoot,
  version,
  resolveNodeModuleDir,
  UnknownProjectError,
} from '@react-native-community/cli-tools';
import findDependencies from './findDependencies';
import resolveReactNativePath from './resolveReactNativePath';
import {
  readConfigFromDisk,
  readDependencyConfigFromDisk,
} from './readConfigFromDisk';
import assign from './assign';
import merge from './merge';

function getDependencyConfig(
  root: string,
  dependencyName: string,
  finalConfig: Config,
  config: UserDependencyConfig,
  userConfig: UserConfig,
  isPlatform: boolean,
): DependencyConfig {
  return merge(
    {
      root,
      name: dependencyName,
      platforms: Object.keys(finalConfig.platforms).reduce(
        (dependency, platform) => {
          const platformConfig = finalConfig.platforms[platform];
          dependency[platform] =
            // Linking platforms is not supported
            isPlatform || !platformConfig
              ? null
              : platformConfig.dependencyConfig(
                  root,
                  config.dependency.platforms[platform],
                );
          return dependency;
        },
        {} as Config['platforms'],
      ),
    },
    userConfig.dependencies[dependencyName] || {},
  ) as DependencyConfig;
}

/**
 * Loads CLI configuration
 */
function loadConfig(projectRoot: string = findProjectRoot()): Config {
  let lazyProject: ProjectConfig;
  const userConfig = readConfigFromDisk(projectRoot);

  const initialConfig: Config = {
    root: projectRoot,
    get reactNativePath() {
      return userConfig.reactNativePath
        ? path.resolve(projectRoot, userConfig.reactNativePath)
        : resolveReactNativePath(projectRoot);
    },
    reactNativeVersion: 'unknown',
    dependencies: userConfig.dependencies,
    commands: userConfig.commands,
    healthChecks: [],
    platforms: userConfig.platforms,
    get project() {
      if (lazyProject) {
        return lazyProject;
      }

      lazyProject = {};
      for (const platform in finalConfig.platforms) {
        const platformConfig = finalConfig.platforms[platform];
        if (platformConfig) {
          lazyProject[platform] = platformConfig.projectConfig(
            projectRoot,
            userConfig.project[platform] || {},
          );
        }
      }

      return lazyProject;
    },
  };

  // Try our best to figure out what version of React Native we're running. This is
  // currently being used to get our deeplinks working, so it's only worried with
  // the major and minor version.
  try {
    let semver = version.current(initialConfig.reactNativePath);
    if (semver) {
      // Retain only these version, since they correspond with our documentation.
      initialConfig.reactNativeVersion = `${semver.major}.${semver.minor}`;
    }
  } catch (e) {
    // If we don't seem to be in a well formed project, give up quietly.
    if (!(e instanceof UnknownProjectError)) {
      throw e;
    }
  }

  const finalConfig = Array.from(
    new Set([
      ...Object.keys(userConfig.dependencies),
      ...findDependencies(projectRoot),
    ]),
  ).reduce((acc: Config, dependencyName) => {
    const localDependencyRoot =
      userConfig.dependencies[dependencyName] &&
      userConfig.dependencies[dependencyName].root;

    try {
      let root =
        localDependencyRoot ||
        resolveNodeModuleDir(projectRoot, dependencyName);
      let config = readDependencyConfigFromDisk(root, dependencyName);

      const isPlatform = Object.keys(config.platforms).length > 0;

      return assign({}, acc, {
        dependencies: assign({}, acc.dependencies, {
          get [dependencyName](): DependencyConfig {
            return getDependencyConfig(
              root,
              dependencyName,
              finalConfig,
              config,
              userConfig,
              isPlatform,
            );
          },
        }),
        commands: [...acc.commands, ...config.commands],
        platforms: {
          ...acc.platforms,
          ...config.platforms,
        },
        healthChecks: [...acc.healthChecks, ...config.healthChecks],
      }) as Config;
    } catch {
      return acc;
    }
  }, initialConfig);

  return finalConfig;
}

export default loadConfig;
