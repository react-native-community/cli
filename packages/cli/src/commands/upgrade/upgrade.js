// @flow
import path from 'path';
import fs from 'fs';
import chalk from 'chalk';
import semver from 'semver';
import execa from 'execa';
import type {ContextT} from '../../tools/types.flow';
import logger from '../../tools/logger';
import PackageManager from '../../tools/PackageManager';
import {fetch} from './helpers';
import legacyUpgrade from './legacyUpgrade';

type FlagsT = {
  legacy: boolean,
};

const rnDiffPurgeUrl =
  'https://github.com/react-native-community/rn-diff-purge';

const getLatestRNVersion = async (): Promise<string> => {
  logger.info('No version passed. Fetching latest...');
  const {stdout} = await execa('npm', ['info', 'react-native', 'version']);
  return stdout;
};

const getRNPeerDeps = async (
  version: string,
): Promise<{[key: string]: string}> => {
  const {stdout} = await execa('npm', [
    'info',
    `react-native@${version}`,
    'peerDependencies',
    '--json',
  ]);

  return JSON.parse(stdout);
};

const getPatch = async (currentVersion, newVersion, projectDir) => {
  let patch;

  const rnDiffAppName = 'RnDiffApp';
  const {name} = require(path.join(projectDir, 'package.json'));

  logger.info(`Fetching diff between v${currentVersion} and v${newVersion}...`);

  try {
    patch = await fetch(
      `${rnDiffPurgeUrl}/compare/version/${currentVersion}...version/${newVersion}.diff`,
    );
  } catch (error) {
    logger.error(
      `Failed to fetch diff for react-native@${newVersion}. Maybe it's not released yet?`,
    );
    logger.info(
      'For available releases to diff see: https://github.com/react-native-community/rn-diff-purge#version-changes',
    );
    return null;
  }

  return patch
    .replace(new RegExp(rnDiffAppName, 'g'), name)
    .replace(new RegExp(rnDiffAppName.toLowerCase(), 'g'), name.toLowerCase());
};

const getVersionToUpgradeTo = async (argv, currentVersion, projectDir) => {
  const newVersion = argv[0]
    ? semver.valid(argv[0]) ||
      (semver.coerce(argv[0]) ? semver.coerce(argv[0]).version : null)
    : await getLatestRNVersion();

  if (!newVersion) {
    logger.error(
      `Provided version "${
        argv[0]
      }" is not allowed. Please pass a valid semver version`,
    );
    return null;
  }

  if (currentVersion > newVersion) {
    logger.error(
      `Trying to upgrade from newer version "${currentVersion}" to older "${newVersion}"`,
    );
    return null;
  }
  if (currentVersion === newVersion) {
    const {
      dependencies: {'react-native': version},
    } = require(path.join(projectDir, 'package.json'));

    if (semver.satisfies(newVersion, version)) {
      logger.warn(
        `Specified version "${newVersion}" is already installed in node_modules and it satisfies "${version}" semver range. No need to upgrade`,
      );
      return null;
    }
    logger.error(
      `Dependency mismatch. Specified version "${newVersion}" is already installed in node_modules and it doesn't satisfy "${version}" semver range of your "react-native" dependency. Please re-install your dependencies`,
    );
    return null;
  }

  return newVersion;
};

const installDeps = async (newVersion, projectDir, patchSuccess) => {
  if (!patchSuccess) {
    logger.warn(
      'Continuing after failure. Most of the files are upgraded but you will need to deal with some conflicts manually',
    );
  }
  logger.info(
    `Installing react-native@${newVersion} and its peer dependencies...`,
  );
  const peerDeps = await getRNPeerDeps(newVersion);
  const pm = new PackageManager({projectDir});
  const deps = [
    `react-native@${newVersion}`,
    ...Object.keys(peerDeps).map(module => `${module}@${peerDeps[module]}`),
  ];
  await pm.install(deps, {silent: true});
  await execa('git', ['add', 'package.json']);
  try {
    await execa('git', ['add', 'yarn.lock']);
  } catch (error) {
    // ignore
  }
  try {
    await execa('git', ['add', 'package-lock.json']);
  } catch (error) {
    // ignore
  }
};

const applyPatch = async (
  currentVersion: string,
  newVersion: string,
  tmpPatchFile: string,
) => {
  let filesToExclude = ['package.json'];
  try {
    try {
      const excludes = filesToExclude.map(e => `--exclude=${e}`);
      await execa('git', [
        'apply',
        '--check',
        tmpPatchFile,
        ...excludes,
        '-p2',
        '--3way',
      ]);
      logger.info('Applying diff...');
    } catch (error) {
      filesToExclude = [
        ...filesToExclude,
        ...error.stderr
          .split('\n')
          .filter(x => x.includes('does not exist in index'))
          .map(x => x.replace(/^error: (.*): does not exist in index$/, '$1')),
      ].filter(Boolean);

      logger.info(`Applying diff (excluding: ${filesToExclude.join(', ')})...`);
    } finally {
      const excludes = filesToExclude.map(e => `--exclude=${e}`);
      await execa('git', ['apply', tmpPatchFile, ...excludes, '-p2', '--3way']);
    }
  } catch (error) {
    if (error.stderr) {
      logger.log(`${chalk.dim(error.stderr.trim())}`);
    }
    logger.error('Automatically applying diff failed');
    logger.info(
      `Here's the diff we tried to apply: ${rnDiffPurgeUrl}/compare/version/${currentVersion}...version/${newVersion}`,
    );
    logger.info(
      `You may find release notes helpful: https://github.com/facebook/react-native/releases/tag/v${newVersion}`,
    );
    return false;
  }
  return true;
};

/**
 * Upgrade application to a new version of React Native.
 */
async function upgrade(argv: Array<string>, ctx: ContextT, args: FlagsT) {
  if (args.legacy) {
    return legacyUpgrade.func(argv, ctx);
  }
  const rnDiffGitAddress =
    'https://github.com/react-native-community/rn-diff-purge.git';
  const tmpRemote = 'tmp-rn-diff-purge';
  const tmpPatchFile = 'tmp-upgrade-rn.patch';
  const projectDir = ctx.root;
  const {version: currentVersion} = require(path.join(
    projectDir,
    'node_modules/react-native/package.json',
  ));

  const newVersion = await getVersionToUpgradeTo(
    argv,
    currentVersion,
    projectDir,
  );

  if (!newVersion) {
    return;
  }

  const patch = await getPatch(currentVersion, newVersion, projectDir);

  if (patch === null) {
    return;
  }

  if (patch === '') {
    logger.info('Diff has no changes to apply, proceeding further');
    await installDeps(newVersion, projectDir);
    logger.success(
      `Upgraded React Native to v${newVersion} ðŸŽ‰. Now you can review and commit the changes`,
    );
    return;
  }
  let patchSuccess;

  try {
    fs.writeFileSync(tmpPatchFile, patch);
    await execa('git', ['remote', 'add', tmpRemote, rnDiffGitAddress]);
    await execa('git', ['fetch', '--no-tags', tmpRemote]);
    patchSuccess = await applyPatch(currentVersion, newVersion, tmpPatchFile);
    if (!patchSuccess) {
      return;
    }
  } catch (error) {
    throw new Error(error.stderr || error);
  } finally {
    try {
      fs.unlinkSync(tmpPatchFile);
    } catch (e) {
      // ignore
    }
    await installDeps(newVersion, projectDir, patchSuccess);
    logger.info('Running "git status" to check what changed...');
    await execa('git', ['status'], {stdio: 'inherit'});
    await execa('git', ['remote', 'remove', tmpRemote]);

    if (!patchSuccess) {
      logger.warn(
        'Please run "git diff" to review the conflicts and resolve them',
      );
      throw new Error(
        'Upgrade failed. Please see the messages above for details',
      );
    }
  }
  logger.success(
    `Upgraded React Native to v${newVersion} ðŸŽ‰. Now you can review and commit the changes`,
  );
}
const upgradeCommand = {
  name: 'upgrade [version]',
  description:
    "Upgrade your app's template files to the specified or latest npm version using `rn-diff-purge` project. Only valid semver versions are allowed.",
  func: upgrade,
  options: [
    {
      command: '--legacy',
      description:
        "Legacy implementation. Upgrade your app's template files to the latest version; run this after " +
        'updating the react-native version in your package.json and running npm install',
    },
  ],
};
export default upgradeCommand;
