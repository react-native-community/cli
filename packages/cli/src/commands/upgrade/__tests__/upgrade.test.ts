import execa from 'execa';
import path from 'path';
import fs from 'fs';
import snapshotDiff from 'snapshot-diff';
import stripAnsi from 'strip-ansi';
import upgrade from '../upgrade';
import {fetch, logger} from '@react-native-community/cli-tools';
import loadConfig from '../../../tools/config';
import merge from '../../../tools/merge';

jest.mock('https');
jest.mock('fs');
jest.mock('path');
jest.mock('execa');
jest.mock(
  '/project/root/node_modules/react-native/package.json',
  () => ({name: 'react-native', version: '0.57.8'}),
  {virtual: true},
);
jest.mock(
  '/project/root/package.json',
  () => ({name: 'TestApp', dependencies: {'react-native': '^0.57.8'}}),
  {virtual: true},
);
jest.mock(
  '/project/root/NestedApp/node_modules/react-native/package.json',
  () => ({name: 'react-native', version: '0.57.8'}),
  {virtual: true},
);
jest.mock(
  '/project/root/NestedApp/package.json',
  () => ({
    name: 'TestAppNested',
    dependencies: {'react-native': '^0.57.8'},
  }),
  {virtual: true},
);
jest.mock('../../../tools/config');
jest.mock('../../../tools/packageManager', () => ({
  install: args => {
    mockPushLog('$ yarn add', ...args);
  },
}));
jest.mock('@react-native-community/cli-tools', () => ({
  ...jest.requireActual('@react-native-community/cli-tools'),
  fetch: jest.fn(),
  logger: {
    info: jest.fn((...args) => mockPushLog('info', args)),
    error: jest.fn((...args) => mockPushLog('error', args)),
    warn: jest.fn((...args) => mockPushLog('warn', args)),
    success: jest.fn((...args) => mockPushLog('success', args)),
    debug: jest.fn((...args) => mockPushLog('debug', args)),
    log: jest.fn((...args) => mockPushLog(args)),
  },
}));

const mockFetch = (value = '', status = 200) => {
  (fetch as jest.Mock).mockImplementation(() =>
    Promise.resolve({data: value, status}),
  );
};

const mockExecaDefault = (command, args) => {
  mockPushLog('$', 'execa', command, args);
  if (command === 'npm' && args[3] === '--json') {
    return Promise.resolve({stdout: '{"react": "16.6.3"}'});
  }
  if (command === 'git' && args[0] === 'rev-parse') {
    return Promise.resolve({stdout: ''});
  }
  return Promise.resolve({stdout: ''});
};

const mockExecaNested = (command, args) => {
  mockPushLog('$', 'execa', command, args);
  if (command === 'npm' && args[3] === '--json') {
    return Promise.resolve({stdout: '{"react": "16.6.3"}'});
  }
  if (command === 'git' && args[0] === 'rev-parse') {
    return Promise.resolve({stdout: 'NestedApp/'});
  }
  return Promise.resolve({stdout: ''});
};

const currentVersion = '0.57.8';
const newVersion = '0.58.4';
const olderVersion = '0.56.0';
const ctx = loadConfig();

const samplePatch = jest
  .requireActual('fs')
  .readFileSync(path.join(__dirname, './sample.patch'), 'utf8');

let logs = [];
const mockPushLog = (...args) =>
  logs.push(args.map(x => (Array.isArray(x) ? x.join(' ') : x)).join(' '));
const flushOutput = () => stripAnsi(logs.join('\n'));

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
  fs.writeFileSync = jest.fn(filename => mockPushLog('[fs] write', filename));
  fs.unlinkSync = jest.fn((...args) => mockPushLog('[fs] unlink', args));
  logs = [];
  ((execa as unknown) as jest.Mock).mockImplementation(mockExecaDefault);
  Object.defineProperty(process, 'platform', {
    value: 'darwin',
  });
});

afterEach(() => {
  fs.writeFileSync = jest.requireMock('fs').writeFileSync;
  fs.unlinkSync = jest.requireMock('fs').unlinkSync;
});

test('uses latest version of react-native when none passed', async () => {
  await upgrade.func([], ctx);
  expect(execa).toBeCalledWith('npm', ['info', 'react-native', 'version']);
}, 60000);

test('applies patch in current working directory when nested', async () => {
  mockFetch(samplePatch, 200);
  ((execa as unknown) as jest.Mock).mockImplementation(mockExecaNested);
  const config = {...ctx, root: '/project/root/NestedApp'};
  await upgrade.func([newVersion], config);

  expect(execa).toBeCalledWith('git', [
    'apply',
    'tmp-upgrade-rn.patch',
    '--exclude=NestedApp/package.json',
    '-p2',
    '--3way',
    '--directory=NestedApp/',
  ]);
});

test('errors when invalid version passed', async () => {
  await upgrade.func(['next'], ctx);
  expect(logger.error).toBeCalledWith(
    'Provided version "next" is not allowed. Please pass a valid semver version',
  );
}, 60000);

test('errors when older version passed', async () => {
  await upgrade.func([olderVersion], ctx);
  expect(logger.error).toBeCalledWith(
    `Trying to upgrade from newer version "${currentVersion}" to older "${olderVersion}"`,
  );
  await upgrade.func(['0.57.10'], ctx);
  expect(logger.error).not.toBeCalledWith(
    `Trying to upgrade from newer version "${currentVersion}" to older "0.57.10"`,
  );
}, 60000);

test('warns when dependency upgrade version is in semver range', async () => {
  await upgrade.func([currentVersion], ctx);
  expect(logger.warn).toBeCalledWith(
    `Specified version "${currentVersion}" is already installed in node_modules and it satisfies "^0.57.8" semver range. No need to upgrade`,
  );
}, 60000);

test('fetches empty patch and installs deps', async () => {
  mockFetch();
  await upgrade.func([newVersion], ctx);
  expect(flushOutput()).toMatchInlineSnapshot(`
    "info Fetching diff between v0.57.8 and v0.58.4...
    info Diff has no changes to apply, proceeding further
    info Installing \\"react-native@0.58.4\\" and its peer dependencies...
    $ execa npm info react-native@0.58.4 peerDependencies --json
    $ yarn add react-native@0.58.4 react@16.6.3
    $ execa git add package.json
    $ execa git add yarn.lock
    $ execa git add package-lock.json
    info Installing CocoaPods dependencies (this may take a few minutes)
    success Upgraded React Native to v0.58.4 ðŸŽ‰. Now you can review and commit the changes"
  `);
}, 60000);

test('fetches regular patch, adds remote, applies patch, installs deps, removes remote,', async () => {
  mockFetch(samplePatch, 200);
  await upgrade.func(
    [newVersion],
    merge(ctx, {
      project: {
        ios: {projectName: 'TestApp.xcodeproj'},
        android: {packageName: 'com.testapp'},
      },
    }),
  );
  expect(flushOutput()).toMatchInlineSnapshot(`
    "info Fetching diff between v0.57.8 and v0.58.4...
    [fs] write tmp-upgrade-rn.patch
    $ execa git rev-parse --show-prefix
    $ execa git apply --binary --check tmp-upgrade-rn.patch --exclude=package.json -p2 --3way --directory=
    info Applying diff...
    $ execa git apply tmp-upgrade-rn.patch --exclude=package.json -p2 --3way --directory=
    [fs] unlink tmp-upgrade-rn.patch
    $ execa git status -s
    info Installing \\"react-native@0.58.4\\" and its peer dependencies...
    $ execa npm info react-native@0.58.4 peerDependencies --json
    $ yarn add react-native@0.58.4 react@16.6.3
    $ execa git add package.json
    $ execa git add yarn.lock
    $ execa git add package-lock.json
    info Installing CocoaPods dependencies (this may take a few minutes)
    info Running \\"git status\\" to check what changed...
    $ execa git status
    success Upgraded React Native to v0.58.4 ðŸŽ‰. Now you can review and commit the changes"
    `);
  expect(
    snapshotDiff(
      samplePatch,
      (fs.writeFileSync as jest.Mock).mock.calls[0][1],
      {
        contextLines: 1,
      },
    ),
  ).toMatchSnapshot(
    'RnDiffApp is replaced with app name (TestApp and com.testapp)',
  );
}, 60000);
test('fetches regular patch, adds remote, applies patch, installs deps, removes remote when updated from nested directory', async () => {
  mockFetch(samplePatch, 200);
  ((execa as unknown) as jest.Mock).mockImplementation(mockExecaNested);
  const config = {...ctx, root: '/project/root/NestedApp'};
  await upgrade.func([newVersion], config);
  expect(flushOutput()).toMatchInlineSnapshot(`
    "info Fetching diff between v0.57.8 and v0.58.4...
    [fs] write tmp-upgrade-rn.patch
    $ execa git rev-parse --show-prefix
    $ execa git apply --binary --check tmp-upgrade-rn.patch --exclude=NestedApp/package.json -p2 --3way --directory=NestedApp/
    info Applying diff...
    $ execa git apply tmp-upgrade-rn.patch --exclude=NestedApp/package.json -p2 --3way --directory=NestedApp/
    [fs] unlink tmp-upgrade-rn.patch
    $ execa git status -s
    info Installing \\"react-native@0.58.4\\" and its peer dependencies...
    $ execa npm info react-native@0.58.4 peerDependencies --json
    $ yarn add react-native@0.58.4 react@16.6.3
    $ execa git add package.json
    $ execa git add yarn.lock
    $ execa git add package-lock.json
    info Installing CocoaPods dependencies (this may take a few minutes)
    info Running \\"git status\\" to check what changed...
    $ execa git status
    success Upgraded React Native to v0.58.4 ðŸŽ‰. Now you can review and commit the changes"
  `);
}, 60000);
test('cleans up if patching fails,', async () => {
  mockFetch(samplePatch, 200);
  ((execa as unknown) as jest.Mock).mockImplementation((command, args) => {
    mockPushLog('$', 'execa', command, args);
    if (command === 'npm' && args[3] === '--json') {
      return Promise.resolve({
        stdout: '{"react": "16.6.3"}',
      });
    }
    if (command === 'git' && args[0] === 'apply') {
      return Promise.reject({
        code: 1,
        stderr:
          'error: .flowconfig: does not exist in index\nerror: ios/MyApp.xcodeproj/project.pbxproj: patch does not apply',
      });
    }
    if (command === 'git' && args[0] === 'rev-parse') {
      return Promise.resolve({stdout: ''});
    }
    return Promise.resolve({stdout: ''});
  });
  try {
    await upgrade.func([newVersion], ctx);
  } catch (error) {
    expect(error.message).toBe(
      'Upgrade failed. Please see the messages above for details',
    );
  }
  expect(flushOutput()).toMatchInlineSnapshot(`
    "info Fetching diff between v0.57.8 and v0.58.4...
    [fs] write tmp-upgrade-rn.patch
    $ execa git rev-parse --show-prefix
    $ execa git apply --binary --check tmp-upgrade-rn.patch --exclude=package.json -p2 --3way --directory=
    info Applying diff...
    warn Excluding files that exist in the template, but not in your project:
      - .flowconfig
    error Excluding files that failed to apply the diff:
      - ios/MyApp.xcodeproj/project.pbxproj
    Please make sure to check the actual changes after the upgrade command is finished.
    You can find them in our Upgrade Helper web app: https://react-native-community.github.io/upgrade-helper/?from=0.57.8&to=0.58.4
    $ execa git apply tmp-upgrade-rn.patch --exclude=package.json --exclude=.flowconfig --exclude=ios/MyApp.xcodeproj/project.pbxproj -p2 --3way --directory=
    debug \\"git apply\\" failed. Error output:
    error: .flowconfig: does not exist in index
    error: ios/MyApp.xcodeproj/project.pbxproj: patch does not apply
    error Automatically applying diff failed. We did our best to automatically upgrade as many files as possible
    [fs] unlink tmp-upgrade-rn.patch
    $ execa git status -s
    error Patch failed to apply for unknown reason. Please fall back to manual way of upgrading
    warn After resolving conflicts don't forget to run \\"pod install\\" inside \\"ios\\" directory
    info You may find these resources helpful:
    â€¢ Release notes: https://github.com/facebook/react-native/releases/tag/v0.58.4
    â€¢ Manual Upgrade Helper: https://react-native-community.github.io/upgrade-helper/?from=0.57.8&to=0.58.4
    â€¢ Git diff: https://raw.githubusercontent.com/react-native-community/rn-diff-purge/diffs/diffs/0.57.8..0.58.4.diff"
  `);
}, 60000);
test('works with --name-ios and --name-android', async () => {
  mockFetch(samplePatch, 200);
  await upgrade.func(
    [newVersion],
    merge(ctx, {
      project: {
        ios: {projectName: 'CustomIos.xcodeproj'},
        android: {packageName: 'co.uk.customandroid.app'},
      },
    }),
  );
  expect(
    snapshotDiff(
      samplePatch,
      (fs.writeFileSync as jest.Mock).mock.calls[0][1],
      {
        contextLines: 1,
      },
    ),
  ).toMatchSnapshot(
    'RnDiffApp is replaced with app name (CustomIos and co.uk.customandroid.app)',
  );
}, 60000);
