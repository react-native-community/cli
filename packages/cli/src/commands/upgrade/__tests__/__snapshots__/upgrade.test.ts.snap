// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches regular patch, adds remote, applies patch, installs deps, removes remote,: RnDiffApp is replaced with app name (TestApp and com.testapp) 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,9 +1,9 @@
- diff --git a/RnDiffApp/android/app/src/main/AndroidManifest.xml b/RnDiffApp/android/app/src/main/AndroidManifest.xml
+ diff --git a/TestApp/android/app/src/main/AndroidManifest.xml b/TestApp/android/app/src/main/AndroidManifest.xml
  index bc3a9310..f3e0d155 100644
- --- a/RnDiffApp/android/app/src/main/AndroidManifest.xml
- +++ b/RnDiffApp/android/app/src/main/AndroidManifest.xml
+ --- a/TestApp/android/app/src/main/AndroidManifest.xml
+ +++ b/TestApp/android/app/src/main/AndroidManifest.xml
  @@ -1,8 +1,7 @@
   <manifest xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
- -    package=\\"com.rndiffapp\\">
- +  package=\\"com.rndiffapp\\">
+ -    package=\\"com.testapp\\">
+ +  package=\\"com.testapp\\">

@@ -14,6 +14,6 @@
         android:name=\\".MainApplication\\"
- diff --git a/RnDiffApp/ios/RnDiffApp/AppDelegate.h b/RnDiffApp/ios/RnDiffApp/AppDelegate.h
+ diff --git a/TestApp/ios/TestApp/AppDelegate.h b/TestApp/ios/TestApp/AppDelegate.h
  index 4b5644f2..2726d5e1 100644
- --- a/RnDiffApp/ios/RnDiffApp/AppDelegate.h
- +++ b/RnDiffApp/ios/RnDiffApp/AppDelegate.h
+ --- a/TestApp/ios/TestApp/AppDelegate.h
+ +++ b/TestApp/ios/TestApp/AppDelegate.h
  @@ -5,9 +5,10 @@
@@ -29,6 +29,6 @@
   @property (nonatomic, strong) UIWindow *window;
- diff --git a/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java b/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java
+ diff --git a/TestApp/android/app/src/main/java/com/testapp/MainApplication.java b/TestApp/android/app/src/main/java/com/testapp/MainApplication.java
  index bc3a9310..f3e0d155 100644
- --- a/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java
- +++ b/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java
+ --- a/TestApp/android/app/src/main/java/com/testapp/MainApplication.java
+ +++ b/TestApp/android/app/src/main/java/com/testapp/MainApplication.java
"
`;

exports[`works with --name-ios and --name-android: RnDiffApp is replaced with app name (CustomIos and co.uk.customandroid.app) 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,9 +1,9 @@
- diff --git a/RnDiffApp/android/app/src/main/AndroidManifest.xml b/RnDiffApp/android/app/src/main/AndroidManifest.xml
+ diff --git a/CustomIos/android/app/src/main/AndroidManifest.xml b/CustomIos/android/app/src/main/AndroidManifest.xml
  index bc3a9310..f3e0d155 100644
- --- a/RnDiffApp/android/app/src/main/AndroidManifest.xml
- +++ b/RnDiffApp/android/app/src/main/AndroidManifest.xml
+ --- a/CustomIos/android/app/src/main/AndroidManifest.xml
+ +++ b/CustomIos/android/app/src/main/AndroidManifest.xml
  @@ -1,8 +1,7 @@
   <manifest xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
- -    package=\\"com.rndiffapp\\">
- +  package=\\"com.rndiffapp\\">
+ -    package=\\"co.uk.customandroid.app\\">
+ +  package=\\"co.uk.customandroid.app\\">

@@ -14,6 +14,6 @@
         android:name=\\".MainApplication\\"
- diff --git a/RnDiffApp/ios/RnDiffApp/AppDelegate.h b/RnDiffApp/ios/RnDiffApp/AppDelegate.h
+ diff --git a/CustomIos/ios/CustomIos/AppDelegate.h b/CustomIos/ios/CustomIos/AppDelegate.h
  index 4b5644f2..2726d5e1 100644
- --- a/RnDiffApp/ios/RnDiffApp/AppDelegate.h
- +++ b/RnDiffApp/ios/RnDiffApp/AppDelegate.h
+ --- a/CustomIos/ios/CustomIos/AppDelegate.h
+ +++ b/CustomIos/ios/CustomIos/AppDelegate.h
  @@ -5,9 +5,10 @@
@@ -29,6 +29,6 @@
   @property (nonatomic, strong) UIWindow *window;
- diff --git a/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java b/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java
+ diff --git a/CustomIos/android/app/src/main/java/co/uk/customandroid/app/MainApplication.java b/CustomIos/android/app/src/main/java/co/uk/customandroid/app/MainApplication.java
  index bc3a9310..f3e0d155 100644
- --- a/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java
- +++ b/RnDiffApp/android/app/src/main/java/com/rndiffapp/MainApplication.java
+ --- a/CustomIos/android/app/src/main/java/co/uk/customandroid/app/MainApplication.java
+ +++ b/CustomIos/android/app/src/main/java/co/uk/customandroid/app/MainApplication.java
"
`;
